### PR TITLE
[WIP] introduce refresh ttl request

### PR DESCRIPTION
**Pull request by birdstorm**
## What have you changed? (mandatory)

Add a refresh lock request so that `TiSpark` and other distributed write service could use it as a solution to long-lasting transactions. This PR makes it possible to write a lock with short TTL and then refresh it at a fixed rate.

## What are the type of the changes? (mandatory)

New feature

## How has this PR been tested? (mandatory)

Tests under construction.

## Does this PR affect documentation (docs) or release note? (mandatory)

No

## Does this PR affect tidb-ansible update? (mandatory)

No

### CODE REVIEW

File: src/server/service/kv.rs [hunk 1]
```diff
@@ -318,6 +318,47 @@ impl<T: RaftStoreRouter + 'static, E: Engine> tikvpb_grpc::Tikv for Service<T, E
...
 
+    fn kv_refresh_lock(
+        &mut self,
+        ctx: RpcContext<'_>,
+        mut req: RefreshLockRequest,
+        sink: UnarySink<RefreshLockResponse>,
+    ) {
+        let timer = GRPC_MSG_HISTOGRAM_VEC.kv_refresh_lock.start_coarse_timer();
+
+        let (cb, f) = paired_future_callback();
```
Comment (youjiali1995) [inline src/server/service/kv.rs]: better to extract a `future_refresh_lock` to be consistent with other functions.

File: src/server/service/kv.rs [hunk 2]
```diff
@@ -318,6 +318,47 @@ impl<T: RaftStoreRouter + 'static, E: Engine> tikvpb_grpc::Tikv for Service<T, E
...
+            cb,
+        );
+        let future = AndThenWith::new(res, f.map_err(Error::from))
+            .and_then(move |v| {
+                let mut resp = RefreshLockResponse::new();
+                if let Some(err) = extract_region_error(&v) {
+                    resp.set_region_error(err);
+                } else {
+                    match v {
+                        Ok(()) => resp.set_ttl(req.get_ttl()),
```
Comment (youjiali1995) [inline src/server/service/kv.rs]: Return the new TTL?

File: src/storage/mvcc/txn.rs [hunk 3]
```diff
@@ -533,6 +533,41 @@ impl<S: Snapshot> MvccTxn<S> {
...
+            if lock.ts != self.start_ts {
+                // locked by another transaction
+                info!(
+                    "locked by another transaction!";
+                    "previous" => self.start_ts,
+                    "current" => lock.ts,
+                );
+
+                return Err(Error::KeyIsLocked {
+                    key: key.to_raw()?,
```
Comment (youjiali1995) [inline src/storage/mvcc/txn.rs]: s/to_raw/into_raw

File: src/storage/mvcc/txn.rs [hunk 4]
```diff
@@ -533,6 +533,41 @@ impl<S: Snapshot> MvccTxn<S> {
...
+                    primary: lock.primary,
+                    ts: lock.ts,
+                    ttl: lock.ttl,
+                    txn_size: options.txn_size,
```
Comment (youjiali1995) [inline src/storage/mvcc/txn.rs]: `txn_size: lock.txn_size`

File: src/storage/mvcc/txn.rs [hunk 5]
```diff
@@ -533,6 +533,41 @@ impl<S: Snapshot> MvccTxn<S> {
...
+                });
+            }
+            // does not consider pessimistic lock for now.
+            self.lock_key(key, lock.lock_type, lock.primary, lock.short_value, options);
```
Comment (youjiali1995) [inline src/storage/mvcc/txn.rs]: Refresh lock should only update lock's TTL and other fields should be unchanged. Please set `options.for_update_ts` and `options.txn_size` to lock's.

File: src/storage/mvcc/txn.rs [hunk 6]
```diff
@@ -533,6 +533,41 @@ impl<S: Snapshot> MvccTxn<S> {
...
+        } else {
+            // the lock is outdated, should abort txn by client.
+            warn!(
+                "lock not found for key";
+                "start_ts" => self.start_ts,
+            );
+            return Err(Error::TxnLockNotFound {
+                start_ts: self.start_ts,
+                commit_ts: u64::max_value(),
+                key: key.as_encoded().to_owned(),
```
Comment (youjiali1995) [inline src/storage/mvcc/txn.rs]: `key: key.into_raw()?`

File: src/storage/mvcc/txn.rs [hunk 7]
```diff
@@ -533,6 +533,41 @@ impl<S: Snapshot> MvccTxn<S> {
...
+            // does not consider pessimistic lock for now.
+            self.lock_key(key, lock.lock_type, lock.primary, lock.short_value, options);
+            Ok(options.lock_ttl)
+        } else {
+            // the lock is outdated, should abort txn by client.
+            warn!(
+                "lock not found for key";
+                "start_ts" => self.start_ts,
+            );
+            return Err(Error::TxnLockNotFound {
```
Comment (youjiali1995) [inline src/storage/mvcc/txn.rs]: remove `return` and `;`

File: src/storage/txn/scheduler.rs [hunk 8]
```diff
@@ -523,6 +523,7 @@ fn gen_command_lock(latches: &Latches, cmd: &Command) -> Lock {
         | Command::PessimisticRollback { ref keys, .. } => latches.gen_lock(keys),
         Command::Cleanup { ref key, .. } => latches.gen_lock(&[key]),
         Command::Pause { ref keys, .. } => latches.gen_lock(keys),
+        Command::RefreshLock { ref key, .. } => latches.gen_lock(&[key]),
```
Comment (youjiali1995) [inline src/storage/txn/scheduler.rs]: What about combining `Command::RefreshLock` with `Command::Cleanup`?

File: src/storage/txn/process.rs [hunk 9]
```diff
@@ -824,6 +823,31 @@ fn process_write_impl<S: Snapshot>(
...
             (ProcessResult::Res, txn.into_modifies(), rows, ctx, None)
         }
+        Command::RefreshLock {
+            ctx,
+            key,
+            start_ts,
+            options,
+        } => {
+            // start_ts + 1 because we want to read a lock with ts = start_ts
+            let mut txn = MvccTxn::new(snapshot.clone(), start_ts + 1, !ctx.get_not_fill_cache())?;
```
Comment (youjiali1995) [inline src/storage/txn/process.rs]: Why + 1?, and `snapshot` need not be cloned.

File: src/storage/mvcc/txn.rs [hunk 10]
```diff
@@ -533,6 +533,41 @@ impl<S: Snapshot> MvccTxn<S> {
...
+                    key: key.to_raw()?,
+                    primary: lock.primary,
+                    ts: lock.ts,
+                    ttl: lock.ttl,
+                    txn_size: options.txn_size,
+                });
+            }
+            // does not consider pessimistic lock for now.
+            self.lock_key(key, lock.lock_type, lock.primary, lock.short_value, options);
+            Ok(options.lock_ttl)
```
Comment (youjiali1995) [inline src/storage/mvcc/txn.rs]: emmm, I think it is useless to return lock's ttl.

File: src/storage/txn/process.rs [hunk 11]
```diff
@@ -824,6 +823,31 @@ fn process_write_impl<S: Snapshot>(
...
+            // may be possible to refresh multiple locks later?
+            let mut locks = vec![];
+            match txn.refresh_lock(key, &options) {
```
Comment (youjiali1995) [inline src/storage/txn/process.rs]: How about
```rust
if let Err(e) = txn.refresh_lock(key, &options) {
    return Err(Error:from(e));
}
statistics.add(&txn.take_statistics());
(ProcessResult::Res, txn.into_modifies(), 1, ctx, None)
```

File: src/storage/mvcc/txn.rs (hunk 10 above)
Comment (MyonKeminta) [reply src/storage/mvcc/txn.rs]: It should be returned through the RPC Response

**Comment by MyonKeminta** [comment]
Please add tests to both `storage/mod.rs` and `txn.rs`.

File: src/storage/mvcc/txn.rs (hunk 10 above)
Comment (youjiali1995) [reply src/storage/mvcc/txn.rs]: I found he didn't use it and we don't have a `StorageCb` that can send back a TTL.😂

Comment (MyonKeminta) [reply src/storage/mvcc/txn.rs]: Oh, I'm wrong. It's not used.

File: src/storage/txn/process.rs (hunk 9 above)
Comment (MyonKeminta) [reply src/storage/txn/process.rs]: Now when it receives such a request with `start_ts = T`, it actually finds the lock whose `start_ts = T + 1`. I guess this is not what we need.

File: src/server/service/kv.rs (hunk 2 above)
Comment (MyonKeminta) [reply src/server/service/kv.rs]: I think it makes more sense to set the ttl to that returned by `MvccTxn::refresh_lock`. Maybe it's better to add a new StorageCb..

Comment (birdstorm) [reply src/server/service/kv.rs]: good idea.

File: src/server/service/kv.rs (hunk 1 above)
Comment (birdstorm) [reply src/server/service/kv.rs]: To do that, I will modify kvproto first to support batch commands. https://github.com/pingcap/kvproto/pull/414

**Comment by birdstorm** [comment]
According to discussion, currently TiDB cannot resolve locks correctly if Primary Lock’s TTL differs from Secondary Lock’s TTL. I will close this PR until TiDB implements `LargeTxn` logic.

**Pull request closed without merging**